### DISCUSSION TITLE
How to handle long running processes

### DISCUSSION THREAD

**Discussion by dominikgeimer** [saloonphp/saloon#129] [Q&A]
Hello,

first of all thanks for this great package. I have a Laravel app and would like to use Saloon(maybe build an SDK for API), but have a question of understanding or would like to know your approach.

I want to query an API and there is an API Endpoint that needs to be addressed to start an operation. Then you have to address a second API endpoint periodically until you finally get the result that the operation is completed. 

Here is a link to the documentation: [MailStore API](https://help.mailstore.com/en/server/Administration_API_-_Using_the_API) -> see Long Running Processes

What would be the best way to do this with Saalon? These processes can take a few seconds, but also several hours. 

Thanks in advance!

**Comment by bumbummen99** [comment]
I would not do this in the request lifecycle but on a queue. You could create seperate requests and then loop it for current status or timeout the job. For one of my projects i save the last "page" of the API i query and continue in the next job so the job does not have to run for hours. On the frontend you could then just check the job status and wait for a job to change the state to succeeded.

**Comment by juse-less** [comment]
I didn't read through the MailStore API thoroughly, so apologies in advance.

I would send the request to initiate the process/operation in MailStore, and then save the operation identifier somewhere, so that you can use cron to regularly poll the API for operations in progress, and then update the state where you saved it (if needed).
By saving the operation identifier, we can use cron to _only_ query for those that are actually in progress.
When an operation is completed, you can do whatever you need, and remove that identifier from the storage you choose.

This could be file(s), a database schema, an in-memory database like Redis, or other means.
If you're only interested in when it's actually completed, you obviously don't need to save the state/progress of the operation - only handle the saving and removal of the identifier.

I'd use a job to both initiate and check operations in progress.
I.e., regardless if it's initiated from a UI or backend 'trigger', dispatch a job, and then the cron would also dispatch background jobs to check for operations in progress.
In terms of cron, I'd probably check for the existence of operations in progress _first_, to not queue unnecessary jobs onto the queue.
